launcher/lobby/lobby_moc.cpp

```diff
@@ -0,0 +1,408 @@
+#include "StdInc.h"
+#include "main.h"
+#include "lobby_moc.h"
+#include "ui_lobby_moc.h"
+#include "lobbyroomrequest_moc.h"
+#include "../mainwindow_moc.h"
+#include "../modManager/cmodlist.h"
+#include "../../lib/CConfigHandler.h"
+
+Lobby::Lobby(QWidget *parent) :
+	QWidget(parent),
+	ui(new Ui::Lobby)
+{
+	ui->setupUi(this);
+
+	connect(&socketLobby, SIGNAL(text(QString)), this, SLOT(sysMessage(QString)));
+	connect(&socketLobby, SIGNAL(receive(QString)), this, SLOT(dispatchMessage(QString)));
+	connect(&socketLobby, SIGNAL(disconnect()), this, SLOT(onDisconnected()));
+	
+	QString hostString("%1:%2");
+	hostString = hostString.arg(QString::fromStdString(settings["launcher"]["lobbyUrl"].String()));
+	hostString = hostString.arg(settings["launcher"]["lobbyPort"].Integer());
+	
+	ui->serverEdit->setText(hostString);
+	ui->userEdit->setText(QString::fromStdString(settings["launcher"]["lobbyUsername"].String()));
+	ui->kickButton->setVisible(false);
+}
+
+Lobby::~Lobby()
+{
+	delete ui;
+}
+
+QMap<QString, QString> Lobby::buildModsMap() const
+{
+	QMap<QString, QString> result;
+	QObject * mainWindow = qApp->activeWindow();
+	if(!mainWindow)
+		mainWindow = parent();
+	if(!mainWindow)
+		return result; //probably something is really wrong here
+	
+	while(mainWindow->parent())
+		mainWindow = mainWindow->parent();
+	const auto & modlist = qobject_cast<MainWindow*>(mainWindow)->getModList();
+	
+	for(auto & modname : modlist.getModList())
+	{
+		auto mod = modlist.getMod(modname);
+		if(mod.isEnabled())
+		{
+			result[modname] = mod.getValue("version").toString();
+		}
+	}
+	return result;
+}
+
+bool Lobby::isModAvailable(const QString & modName, const QString & modVersion) const
+{
+	QObject * mainWindow = qApp->activeWindow();
+	while(mainWindow->parent())
+		mainWindow = mainWindow->parent();
+	const auto & modlist = qobject_cast<MainWindow*>(mainWindow)->getModList();
+	
+	if(!modlist.hasMod(modName))
+		return false;
+
+	auto mod = modlist.getMod(modName);
+	return (mod.isInstalled () || mod.isAvailable()) && (mod.getValue("version") == modVersion);
+}
+
+void Lobby::serverCommand(const ServerCommand & command) try
+{
+	//initialize variables outside of switch block
+	const QString statusPlaceholder("%1 %2\n");
+	const auto & args = command.arguments;
+	int amount, tagPoint;
+	QString joinStr;
+	switch(command.command)
+	{
+	case SRVERROR:
+		protocolAssert(args.size());
+		chatMessage("System error", args[0], true);
+		if(authentificationStatus == AuthStatus::AUTH_NONE)
+			authentificationStatus = AuthStatus::AUTH_ERROR;
+		break;
+
+	case CREATED:
+		protocolAssert(args.size());
+		hostSession = args[0];
+		session = args[0];
```

Comment:
both args[0]? Is it because it is host?

Reply:
It means that you are hosting session and you are joined to the same session

Reply:
Are command args explained/documented somewhere?
args[0] is not very descriptive name.
Question does for all commands, not just CREATED

Reply:
In the header there is description for server commands and their arguments